Unified to throw NotSupportedException when SendFile() for connectionless sockets

rhirano0715 opened the pull request:
The issue was that the Socket.SendFile() threw inconsistent exceptions when the platform was Windows and the protocol was UDP. The first call would throw a SocketException, while the second call would throw a NotSupportedException.

With this commit, SendFile() will consistently throw NotSupportException on all platforms when the protocol is UDP.

Fix #47472

rhirano0715 committed:
Unified to throw NotSupportedException when SendFile() for connectionless sockets

The issue was that the Socket.SendFile() threw inconsistent exceptions when the platform was Windows and the protocol was UDP.
The first call would throw a SocketException, while the second call would throw a NotSupportedException.

With this commit, SendFile() will consistently throw NotSupportException on all platforms when the protocol is UDP.

Fix #47472

rhirano0715 commented:
@dotnet-policy-service agree

karelz commented:
@antonfirsov ping, can you please take a look?

antonfirsov commented on src/libraries/System.Net.Sockets/src/System/Net/Sockets/Socket.Tasks.cs:
```diff
@@ -670,6 +670,11 @@ public ValueTask SendFileAsync(string? fileName, CancellationToken cancellationT
         /// <exception cref="SocketException">An error occurred when attempting to access the socket.</exception>
         public ValueTask SendFileAsync(string? fileName, ReadOnlyMemory<byte> preBuffer, ReadOnlyMemory<byte> postBuffer, TransmitFileOptions flags, CancellationToken cancellationToken = default)
         {
+            if (_protocolType == ProtocolType.Udp)
+            {
+                throw new NotSupportedException(SR.net_notconnected);
+            }
```
Doesn't the check `if (!IsConnectionOriented)` do the job? In #47472 I only tested the sync overload.

antonfirsov commented on src/libraries/System.Net.Sockets/src/System/Net/Sockets/Socket.cs:
```diff
@@ -1254,6 +1254,11 @@ public void SendFile(string? fileName, byte[]? preBuffer, byte[]? postBuffer, Tr
         /// <exception cref="SocketException">An error occurred when attempting to access the socket.</exception>
         public void SendFile(string? fileName, ReadOnlySpan<byte> preBuffer, ReadOnlySpan<byte> postBuffer, TransmitFileOptions flags)
         {
+            if (_protocolType == ProtocolType.Udp)
+            {
+                throw new NotSupportedException(SR.net_notconnected);
+            }
```
The async variant checks for `if (!IsConnectionOriented)` while the sync version checks `(!Connected)`. I think that - instead of limiting the error case explicitly to UDP - we should consolidate these checks, throwing in case of `(!IsConnectionOriented || !Connected)`.

antonfirsov reviewed (commented):
Sorry for the long delay!

karelz commented:
@rhirano0715 will you have time to look at the CR feedback from @antonfirsov please?

rhirano0715 committed:
Merge branch 'dotnet:main' into issue47472

rhirano0715 committed:
Change to throws `NotSupportedException` if `!IsConnectionOriented` or `!Connected`.

Before:.

    Throws `NotSupportedException` on UDP.

After:

    Throws `NotSupportedException` if `!IsConnectionOriented` or `!Connected`.

rhirano0715 committed:
Changed test case `UdpConnection_ThrowsException` to run regardless of platform.

rhirano0715 commented:
@karelz   @antonfirsov 
I'm sorry for not being able to respond.
Thank you for your review.
I have committed the correction.

karelz commented on src/libraries/System.Net.Sockets/src/System/Net/Sockets/Socket.Tasks.cs:
```diff
@@ -677,8 +677,8 @@ public ValueTask SendFileAsync(string? fileName, ReadOnlyMemory<byte> preBuffer,
             {
-                var soex = new SocketException((int)SocketError.NotConnected);
-                return ValueTask.FromException(soex);
+                var nsex = new NotSupportedException(SR.net_notconnected);
+                return ValueTask.FromException(nsex);
```
```suggestion
                var ex = new NotSupportedException(SR.net_notconnected);
                return ValueTask.FromException(ex);
```

rhirano0715 committed:
Update src/libraries/System.Net.Sockets/src/System/Net/Sockets/Socket.Tasks.cs

Co-authored-by: Karel Zikmund <karelz@microsoft.com>

antonfirsov commented:
/azp run runtime

antonfirsov commented:
/azp run runtime-extra-platforms

antonfirsov reviewed (approved):
LGTM, thanks, and sorry for the slow response. Started some extra CI runs just in case.

MihaZupan commented:
/azp run runtime-extra-platforms
